src/Tools/ExternalAccess/Razor/Cohost/RazorDynamicRegistrationServiceFactory.cs

```diff
@@ -44,17 +46,29 @@ public RazorDynamicRegistrationService(IRazorCohostDynamicRegistrationService? d
 
         public Task OnInitializedAsync(ClientCapabilities clientCapabilities, RequestContext context, CancellationToken cancellationToken)
         {
-            if (_dynamicRegistrationService is null)
+            if (_dynamicRegistrationService is null || _clientLanguageServerManager is null)
             {
                 return Task.CompletedTask;
             }
 
+            var uiContext = UIContext.FromUIContextGuid(Constants.RazorCohostingUIContext);
+            uiContext.WhenActivated(() =>
+            {
+                // Not using the cancellation token passed in, as the context could be activated well after LSP server initialization
```

Comment:
Hmm, might need to expose a way for something to access a general server cancellation token for things that should be cancelled when the server shutsdown.

Reply:
Thats a good idea

Reply:
`IDisposable` was enough for this, and much simpler.

One thing I noticed from my testing though, as FYI, the "AlwaysActive" server initializes when Roslyn loads, and then shuts down when a solution is closed. Fine, makes sense. BUT if I then open a solution again, it doesn't get initialized again until a C# file is opened. In other words. No idea if that could/would be a problem ever.

Reply:
Right, dispose should take care of it since its an LSP service - good call.

> BUT if I then open a solution again, it doesn't get initialized again until a C# file is opened. In other words. No idea if that could/would be a problem ever.

Yes, theres a potential bug there - workspace level requests won't get fulfilled on the second solution open unti la file is opened.  I think its because the VS workspace doesn't get recreated (so the event listener doesn't trigger).